### PR TITLE
21 email login + 22 anonymous login

### LINKED ISSUE #21: Add sign in with email and password with the Firebase backend to the login page

### LINKED ISSUE #22: Add anonymous sign in with the Firebase backend to the login page

### PR DESCRIPTION
**Pull request by clementcharmillot**
- Adds the possibility to register with an email address.
- Adds the possibility to register with an email address.
- Adds the possibility to register anonymously.

Limitations:
- Registration doesn't redirect to the account registration screen. Waiting for #23.

closes #21 
closes #22

### CODE REVIEW

File: app/src/main/java/com/epfl/drawyourpath/authentication/FirebaseAuth.kt
Comment (damiengenoud) [inline app/src/main/java/com/epfl/drawyourpath/authentication/FirebaseAuth.kt:1]: Adding comments for this type of classes could be very helpful for better understanding of what is going on. (Though I might not be the most qualified member on the login and authentication topic.)

File: app/src/main/java/com/epfl/drawyourpath/login/LoginActivity.kt
Comment (damiengenoud) [inline app/src/main/java/com/epfl/drawyourpath/login/LoginActivity.kt:1]: Adding comments for this type of classes could be very helpful for better understanding of what is going on. (Though I might not be the most qualified member on the login and authentication topic.)

**Review by damiengenoud — COMMENTED** [review]
Very good PR, nothing to say on the code itself except that it is pretty hard to follow everything so some comments would be great.

I do have some questions: when logging in with Google (after registration), is it normal that I get an exception ? When trying to not use an account (logging in anonymously) nothing happens, is it normal ?

I think adding a complete description of what the PR implements and what is not done yet could help answering those questions.

Also very good test class, it looks like you covered many possible failure cases.
Good job and thank you for all this work!

**Comment by clementcharmillot** [comment]
@damiengenoud Thanks for the review. 

Just tried Google authentication and didn't trigger any exception. Could you please share the reproduction steps?

After registration it stays on the login activity as we need to redirect to the registration activity which isn't yet available. (#23)

You are right, it really lacks comments. I'll add some tomorrow.

**Comment by damiengenoud** [comment]
Capture d’écran 2023-03-15 025849

The "exception" I'm talking about is this one, which is displayed whenever I register or login with Google.
Register: it happens after I connect to my account, so instantly when I'm back in the register screen.
Login: it happens  after I click on my account when prompted to choose one, so instantly when I'm back in the login screen.

I don't know if there is something missing on my end or if it is a normal behaviour (for now).

**Comment by clementcharmillot** [comment]
@damiengenoud I wiped out the data on my emulator and tried again but didn't get any exception. The code you are getting is caused by an invalid SHA1 key in the Firebase Console. As I really don't know how the SHA1 key works, it might simply mean that it isn't similar on your device. 

You can generate the key with the `gradle signingreport` command. If you send it privately I can add it to the Firebase Console.

**Comment by clementcharmillot** [comment]
Last commit also includes the link between login, main menu and registration form.

**Comment by damiengenoud** [comment]
Thank you for adding the comment, it really helps understanding what is going on !

**Review by damiengenoud — APPROVED** [review]
Thank you for the few changes, this looks fine to me.
